bus-mapping/src/circuit_input_builder.rs

```diff
@@ -869,31 +1086,65 @@ impl<'a> CircuitInputBuilder {
         eth_tx: &eth_types::Transaction,
         geth_trace: &GethExecTrace,
     ) -> Result<(), Error> {
-        let mut tx = self.new_tx(eth_tx)?;
+        let mut call_is_success = HashMap::new();
+        let mut call_indices = Vec::new();
+        for (index, geth_step) in geth_trace.struct_logs.iter().enumerate() {
+            if let Some(geth_next_step) = geth_trace.struct_logs.get(index + 1)
```

Comment:
It makes sense to me, I would like to propose to add a method `prepare` of `CircuitInputBuilder`, which takes `(client, txs, geth_traces)` to iterate over `geth_traces` and do:

1. Collect access trace (`gen_state_access_trace`) to get required `(accounts, codes, proofs)` by `client`
2. Collect meta information of a call like `is_success`

The benefit is, we can keep `CircuitInputBuilder`'s lifetime longer and build witness incrementally. After each input of a block is generated, we can keep the `state_db` and `code_db` and `prepare` next block.

Also I think we can do it later, otherwise I will be trapped in such rebasing hell (my apologies to touch to many files).

Reply:
Indeed, a followup PR might suit better! :)

Reply:
I also thought about keeping the CircuitInputBuilder from one block to the next, so that we are able to keep some values from the state_db and code_db and possibly reduce the number of rpc queries to geth.  It seems we all agree on working on this change :)

I agree that we can move this task to a future PR, so I'll resolve the conversation and open an issue regarding this change.